`<expected>`: `std::function` returning `expected<any, int>` emits error C7608: atomic constraint should be a constant expression

Originally reported by Jared Resch as https://godbolt.org/z/YTe7hYPrq . I've further reduced this to https://godbolt.org/z/Ebadxz43K which is accepted by libstdc++ and libc++, but rejected by VS 2022 17.8 Preview 1:

```
C:\Temp>type meow.cpp
```
```cpp
#include <any>
#include <expected>
#include <functional>
using namespace std;

using X = expected<any, int>;

X meow() {
    return {};
}

int main() {
    function<X()> f{meow};
}
```
```
C:\Temp>cl /EHsc /nologo /W4 /std:c++latest meow.cpp
meow.cpp
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(258): error C7608: atomic constraint should be a constant expression
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(258): note: the template instantiation context (the oldest one first) is
meow.cpp(13): note: while processing the default template argument of 'std::function<X (void)>::function(_Fx &&)'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\functional(1050): note: see reference to alias template instantiation 'std::_Func_class<_Ret>::_Enable_if_callable_t<X(__cdecl &)(void),std::function<X (void)>>' being compiled
        with
        [
            _Ret=X
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\functional(892): note: see reference to variable template 'const bool conjunction_v<std::negation<std::is_same<std::expected<std::any,int> __cdecl(void),std::function<std::expected<std::any,int> __cdecl(void)> > >,std::_Is_invocable_r<std::expected<std::any,int>,std::expected<std::any,int> (__cdecl*&)(void)> >' being compiled
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\functional(892): note: see reference to class template instantiation 'std::_Is_invocable_r<_Ret,X (__cdecl *&)(void)>' being compiled
        with
        [
            _Ret=X
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\type_traits(1889): note: see reference to alias template instantiation 'std::_Is_invocable_r_<_Rx,_Callable,>' being compiled
        with
        [
            _Rx=X,
            _Callable=X (__cdecl *&)(void)
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\type_traits(1886): note: see reference to alias template instantiation 'std::_Invoke_traits_common<std::expected<std::any,int>,false>::_Is_invocable_r<_Rx>' being compiled
        with
        [
            _Rx=X
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\type_traits(1810): note: see reference to variable template 'const bool disjunction_v<std::is_void<std::expected<std::any,int> >,std::_Invoke_convertible<std::expected<std::any,int>,std::expected<std::any,int>,void> >' being compiled
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(271): note: see reference to variable template 'const bool std::expected<std::any,int>::_Allow_unwrapping<std::any,int>' being compiled
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(242): note: see reference to variable template 'const bool disjunction_v<std::is_same<std::any,bool>,std::negation<std::disjunction<std::is_constructible<std::any,std::expected<std::any,int> &>,std::is_constructible<std::any,std::expected<std::any,int> >,std::is_constructible<std::any,std::expected<std::any,int> const &>,std::is_constructible<std::any,std::expected<std::any,int> const >,std::is_convertible<std::expected<std::any,int> &,std::any>,std::is_convertible<std::expected<std::any,int> &&,std::any>,std::is_convertible<std::expected<std::any,int> const &,std::any>,std::is_convertible<std::expected<std::any,int> const &&,std::any> > > >' being compiled
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(250): note: see reference to class template instantiation 'std::is_constructible<_Ty,std::expected<_Ty,int> &>' being compiled
        with
        [
            _Ty=std::any
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\type_traits(755): note: while processing the default template argument of 'std::any::any(_ValueType &&)'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\any(138): note: see reference to variable template 'const bool conjunction_v<std::negation<std::is_same<std::expected<std::any,int>,std::any> >,std::negation<std::_Is_specialization<std::expected<std::any,int>,std::in_place_type_t> >,std::is_copy_constructible<std::expected<std::any,int> > >' being compiled
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(258): error C2131: expression did not evaluate to a constant
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(258): note: failure was caused by a read of an uninitialized symbol
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(258): note: see usage of 'std::expected<std::any,int>::_Allow_unwrapping<std::any,int>'
```

I haven't figured out the root cause yet (library bug? compiler bug? an LWG issue resolution we've implemented that the others haven't?).

--- a/stl/inc/expected
+++ b/stl/inc/expected
@@ -238,6 +238,9 @@ public:
     // clang-format on
 
     template <class _Uty, class _UErr>
+    static constexpr bool _Not_self = !is_same_v<_Ty, _Uty> || !is_same_v<_Err, _UErr>;
+
+    template <class _Uty, class _UErr>
     static constexpr bool _Allow_unwrapping = disjunction_v<is_same<remove_cv_t<_Ty>, bool>,
                                                   negation<disjunction<is_constructible<_Ty, expected<_Uty, _UErr>&>, //
                                                       is_constructible<_Ty, expected<_Uty, _UErr>>, //
@@ -253,8 +256,8 @@ public:
                                            && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>>;
 
     template <class _Uty, class _UErr>
-        requires is_constructible_v<_Ty, const _Uty&> && is_constructible_v<_Err, const _UErr&>
-              && _Allow_unwrapping<_Uty, _UErr>
+        requires _Not_self<_Uty, _UErr> && is_constructible_v<_Ty, const _Uty&>
+              && is_constructible_v<_Err, const _UErr&> && _Allow_unwrapping<_Uty, _UErr>
     constexpr explicit(!is_convertible_v<const _Uty&, _Ty> || !is_convertible_v<const _UErr&, _Err>)
         expected(const expected<_Uty, _UErr>& _Other) noexcept(is_nothrow_constructible_v<_Ty, const _Uty&> //
                 && is_nothrow_constructible_v<_Err, const _UErr&>) // strengthened
@@ -267,7 +270,8 @@ public:
     }
 
     template <class _Uty, class _UErr>
-        requires is_constructible_v<_Ty, _Uty> && is_constructible_v<_Err, _UErr> && _Allow_unwrapping<_Uty, _UErr>
+        requires _Not_self<_Uty, _UErr> && is_constructible_v<_Ty, _Uty> && is_constructible_v<_Err, _UErr>
+              && _Allow_unwrapping<_Uty, _UErr>
     constexpr explicit(!is_convertible_v<_Uty, _Ty> || !is_convertible_v<_UErr, _Err>)
         expected(expected<_Uty, _UErr>&& _Other) noexcept(
             is_nothrow_constructible_v<_Ty, _Uty>&& is_nothrow_constructible_v<_Err, _UErr>) // strengthened
@@ -342,7 +346,7 @@ public:
     }
 
     // clang-format off
-    ~expected() requires is_trivially_destructible_v<_Ty> && is_trivially_destructible_v<_Err> = default;
+    ~expected() requires is_trivially_destructible_v<_Ty>&& is_trivially_destructible_v<_Err> = default;
     // clang-format on
 
     // [expected.object.assign]
@@ -1202,13 +1206,17 @@ public:
     // clang-format on
 
     template <class _Uty, class _UErr>
+    static constexpr bool _Not_self = !is_same_v<_Ty, _Uty> || !is_same_v<_Err, _UErr>;
+
+    template <class _Uty, class _UErr>
     static constexpr bool _Allow_unwrapping = !is_constructible_v<unexpected<_Err>, expected<_Uty, _UErr>&>
                                            && !is_constructible_v<unexpected<_Err>, expected<_Uty, _UErr>>
                                            && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>&>
                                            && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>>;
 
     template <class _Uty, class _UErr>
-        requires is_void_v<_Uty> && is_constructible_v<_Err, const _UErr&> && _Allow_unwrapping<_Uty, _UErr>
+        requires _Not_self<_Uty, _UErr> && is_void_v<_Uty> && is_constructible_v<_Err, const _UErr&>
+              && _Allow_unwrapping<_Uty, _UErr>
     constexpr explicit(!is_convertible_v<const _UErr&, _Err>) expected(const expected<_Uty, _UErr>& _Other) noexcept(
         is_nothrow_constructible_v<_Err, const _UErr&>) // strengthened
         : _Has_value(_Other._Has_value) {
@@ -1218,7 +1226,8 @@ public:
     }
 
     template <class _Uty, class _UErr>
-        requires is_void_v<_Uty> && is_constructible_v<_Err, _UErr> && _Allow_unwrapping<_Uty, _UErr>
+        requires _Not_self<_Uty, _UErr> && is_void_v<_Uty> && is_constructible_v<_Err, _UErr>
+              && _Allow_unwrapping<_Uty, _UErr>
     constexpr explicit(!is_convertible_v<_UErr, _Err>)
         expected(expected<_Uty, _UErr>&& _Other) noexcept(is_nothrow_constructible_v<_Err, _UErr>) // strengthened
         : _Has_value(_Other._Has_value) {

--- a/stl/inc/expected
+++ b/stl/inc/expected
@@ -237,6 +237,7 @@ public:
         is_trivially_move_constructible_v<_Ty> && is_trivially_move_constructible_v<_Err> = default;
     // clang-format on
 
+private:
     template <class _Uty, class _UErr>
     static constexpr bool _Not_self = !is_same_v<_Ty, _Uty> || !is_same_v<_Err, _UErr>;
 
@@ -255,6 +256,7 @@ public:
                                            && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>&> //
                                            && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>>;
 
+public:
     template <class _Uty, class _UErr>
         requires _Not_self<_Uty, _UErr> && is_constructible_v<_Ty, const _Uty&>
               && is_constructible_v<_Err, const _UErr&> && _Allow_unwrapping<_Uty, _UErr>
@@ -1205,6 +1207,7 @@ public:
     expected(expected&&) requires is_trivially_move_constructible_v<_Err> = default;
     // clang-format on
 
+private:
     template <class _Uty, class _UErr>
     static constexpr bool _Not_self = !is_same_v<_Ty, _Uty> || !is_same_v<_Err, _UErr>;
 
@@ -1214,6 +1217,7 @@ public:
                                            && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>&>
                                            && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>>;
 
+public:
     template <class _Uty, class _UErr>
         requires _Not_self<_Uty, _UErr> && is_void_v<_Uty> && is_constructible_v<_Err, const _UErr&>
               && _Allow_unwrapping<_Uty, _UErr>
@@ -1389,26 +1393,7 @@ public:
         is_nothrow_move_constructible_v<_Err>&& is_nothrow_swappable_v<_Err>)
         requires is_swappable_v<_Err> && is_move_constructible_v<_Err>
     {
-        using _STD swap;
-        if (_Left._Has_value && _Right._Has_value) {
-            // nothing
-        } else if (_Left._Has_value) {
-            _STD construct_at(_STD addressof(_Left._Unexpected), _STD move(_Right._Unexpected));
-            if constexpr (!is_trivially_destructible_v<_Err>) {
-                _Right._Unexpected.~_Err();
-            }
-            _Left._Has_value  = false;
-            _Right._Has_value = true;
-        } else if (_Right._Has_value) {
-            _STD construct_at(_STD addressof(_Right._Unexpected), _STD move(_Left._Unexpected));
-            if constexpr (!is_trivially_destructible_v<_Err>) {
-                _Left._Unexpected.~_Err();
-            }
-            _Left._Has_value  = true;
-            _Right._Has_value = false;
-        } else {
-            swap(_Left._Unexpected, _Right._Unexpected); // intentional ADL
-        }
+        _Left.swap(_Right);
     }
 
     // [expected.void.obs]

--- a/stl/inc/expected
+++ b/stl/inc/expected
@@ -190,6 +190,21 @@ class expected {
     template <class _UTy, class _UErr>
     friend class expected;
 
+    template <class _Uty, class _UErr>
+    static constexpr bool _Allow_unwrapping = disjunction_v<is_same<remove_cv_t<_Ty>, bool>,
+                                                  negation<disjunction<is_constructible<_Ty, expected<_Uty, _UErr>&>, //
+                                                      is_constructible<_Ty, expected<_Uty, _UErr>>, //
+                                                      is_constructible<_Ty, const expected<_Uty, _UErr>&>, //
+                                                      is_constructible<_Ty, const expected<_Uty, _UErr>>, //
+                                                      is_convertible<expected<_Uty, _UErr>&, _Ty>, //
+                                                      is_convertible<expected<_Uty, _UErr>&&, _Ty>, //
+                                                      is_convertible<const expected<_Uty, _UErr>&, _Ty>, //
+                                                      is_convertible<const expected<_Uty, _UErr>&&, _Ty>>>> //
+                                           && !is_constructible_v<unexpected<_Err>, expected<_Uty, _UErr>&> //
+                                           && !is_constructible_v<unexpected<_Err>, expected<_Uty, _UErr>> //
+                                           && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>&> //
+                                           && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>>;
+
 public:
     using value_type      = _Ty;
     using error_type      = _Err;
@@ -237,28 +252,8 @@ public:
         is_trivially_move_constructible_v<_Ty> && is_trivially_move_constructible_v<_Err> = default;
     // clang-format on
 
-private:
     template <class _Uty, class _UErr>
-    static constexpr bool _Not_self = !is_same_v<_Ty, _Uty> || !is_same_v<_Err, _UErr>;
-
-    template <class _Uty, class _UErr>
-    static constexpr bool _Allow_unwrapping = disjunction_v<is_same<remove_cv_t<_Ty>, bool>,
-                                                  negation<disjunction<is_constructible<_Ty, expected<_Uty, _UErr>&>, //
-                                                      is_constructible<_Ty, expected<_Uty, _UErr>>, //
-                                                      is_constructible<_Ty, const expected<_Uty, _UErr>&>, //
-                                                      is_constructible<_Ty, const expected<_Uty, _UErr>>, //
-                                                      is_convertible<expected<_Uty, _UErr>&, _Ty>, //
-                                                      is_convertible<expected<_Uty, _UErr>&&, _Ty>, //
-                                                      is_convertible<const expected<_Uty, _UErr>&, _Ty>, //
-                                                      is_convertible<const expected<_Uty, _UErr>&&, _Ty>>>> //
-                                           && !is_constructible_v<unexpected<_Err>, expected<_Uty, _UErr>&> //
-                                           && !is_constructible_v<unexpected<_Err>, expected<_Uty, _UErr>> //
-                                           && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>&> //
-                                           && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>>;
-
-public:
-    template <class _Uty, class _UErr>
-        requires _Not_self<_Uty, _UErr> && is_constructible_v<_Ty, const _Uty&>
+        requires _Different_from<expected<_Uty, _UErr>, expected> && is_constructible_v<_Ty, const _Uty&>
               && is_constructible_v<_Err, const _UErr&> && _Allow_unwrapping<_Uty, _UErr>
     constexpr explicit(!is_convertible_v<const _Uty&, _Ty> || !is_convertible_v<const _UErr&, _Err>)
         expected(const expected<_Uty, _UErr>& _Other) noexcept(is_nothrow_constructible_v<_Ty, const _Uty&> //
@@ -272,8 +267,8 @@ public:
     }
 
     template <class _Uty, class _UErr>
-        requires _Not_self<_Uty, _UErr> && is_constructible_v<_Ty, _Uty> && is_constructible_v<_Err, _UErr>
-              && _Allow_unwrapping<_Uty, _UErr>
+        requires _Different_from<expected<_Uty, _UErr>, expected> && is_constructible_v<_Ty, _Uty>
+              && is_constructible_v<_Err, _UErr> && _Allow_unwrapping<_Uty, _UErr>
     constexpr explicit(!is_convertible_v<_Uty, _Ty> || !is_convertible_v<_UErr, _Err>)
         expected(expected<_Uty, _UErr>&& _Other) noexcept(
             is_nothrow_constructible_v<_Ty, _Uty>&& is_nothrow_constructible_v<_Err, _UErr>) // strengthened
@@ -1172,6 +1167,12 @@ class expected<_Ty, _Err> {
     template <class _UTy, class _UErr>
     friend class expected;
 
+    template <class _Uty, class _UErr>
+    static constexpr bool _Allow_unwrapping = !is_constructible_v<unexpected<_Err>, expected<_Uty, _UErr>&>
+                                           && !is_constructible_v<unexpected<_Err>, expected<_Uty, _UErr>>
+                                           && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>&>
+                                           && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>>;
+
 public:
     using value_type      = _Ty;
     using error_type      = _Err;
@@ -1207,20 +1208,9 @@ public:
     expected(expected&&) requires is_trivially_move_constructible_v<_Err> = default;
     // clang-format on
 
-private:
     template <class _Uty, class _UErr>
-    static constexpr bool _Not_self = !is_same_v<_Ty, _Uty> || !is_same_v<_Err, _UErr>;
-
-    template <class _Uty, class _UErr>
-    static constexpr bool _Allow_unwrapping = !is_constructible_v<unexpected<_Err>, expected<_Uty, _UErr>&>
-                                           && !is_constructible_v<unexpected<_Err>, expected<_Uty, _UErr>>
-                                           && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>&>
-                                           && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>>;
-
-public:
-    template <class _Uty, class _UErr>
-        requires _Not_self<_Uty, _UErr> && is_void_v<_Uty> && is_constructible_v<_Err, const _UErr&>
-              && _Allow_unwrapping<_Uty, _UErr>
+        requires _Different_from<expected<_Uty, _UErr>, expected> && is_void_v<_Uty>
+              && is_constructible_v<_Err, const _UErr&> && _Allow_unwrapping<_Uty, _UErr>
     constexpr explicit(!is_convertible_v<const _UErr&, _Err>) expected(const expected<_Uty, _UErr>& _Other) noexcept(
         is_nothrow_constructible_v<_Err, const _UErr&>) // strengthened
         : _Has_value(_Other._Has_value) {
@@ -1230,7 +1220,7 @@ public:
     }
 
     template <class _Uty, class _UErr>
-        requires _Not_self<_Uty, _UErr> && is_void_v<_Uty> && is_constructible_v<_Err, _UErr>
+        requires _Different_from<expected<_Uty, _UErr>, expected> && is_void_v<_Uty> && is_constructible_v<_Err, _UErr>
               && _Allow_unwrapping<_Uty, _UErr>
     constexpr explicit(!is_convertible_v<_UErr, _Err>)
         expected(expected<_Uty, _UErr>&& _Other) noexcept(is_nothrow_constructible_v<_Err, _UErr>) // strengthened

--- a/stl/inc/expected
+++ b/stl/inc/expected
@@ -346,7 +346,7 @@ public:
     }
 
     // clang-format off
-    ~expected() requires is_trivially_destructible_v<_Ty>&& is_trivially_destructible_v<_Err> = default;
+    ~expected() requires is_trivially_destructible_v<_Ty> && is_trivially_destructible_v<_Err> = default;
     // clang-format on
 
     // [expected.object.assign]

--- a/stl/inc/expected
+++ b/stl/inc/expected
@@ -184,6 +184,7 @@ struct _Check_expected_argument : true_type {
 
 _EXPORT_STD template <class _Ty, class _Err>
 class expected {
+private:
     static_assert(_Check_expected_argument<_Ty>::value);
     static_assert(_Check_unexpected_argument<_Err>::value);
 
@@ -1162,6 +1163,7 @@ private:
 template <class _Ty, class _Err>
     requires is_void_v<_Ty>
 class expected<_Ty, _Err> {
+private:
     static_assert(_Check_unexpected_argument<_Err>::value);
 
     template <class _UTy, class _UErr>

--- a/tests/std/tests/P0323R12_expected/test.cpp
+++ b/tests/std/tests/P0323R12_expected/test.cpp
@@ -2136,10 +2136,9 @@ void test_lwg_3843() {
     }
 }
 
-void test_gh_4011() {
-    static_assert(copyable<expected<any, int>>);
-    static_assert(copyable<expected<void, any>>);
-}
+// Test GH-4011: these predicates triggered constraint recursion.
+static_assert(copyable<expected<any, int>>);
+static_assert(copyable<expected<void, any>>);
 
 int main() {
     test_unexpected::test_all();
@@ -2157,5 +2156,4 @@ int main() {
 
     test_reinit_regression();
     test_lwg_3843();
-    test_gh_4011();
 }

--- a/tests/std/tests/P0323R12_expected/test.cpp
+++ b/tests/std/tests/P0323R12_expected/test.cpp
@@ -3,6 +3,7 @@
 
 #define _CONTAINER_DEBUG_LEVEL 1
 
+#include <any>
 #include <cassert>
 #include <concepts>
 #include <exception>
@@ -2135,6 +2136,11 @@ void test_lwg_3843() {
     }
 }
 
+void test_gh_4011() {
+    static_assert(copyable<expected<any, int>>);
+    static_assert(copyable<expected<void, any>>);
+}
+
 int main() {
     test_unexpected::test_all();
     static_assert(test_unexpected::test_all());
@@ -2151,4 +2157,5 @@ int main() {
 
     test_reinit_regression();
     test_lwg_3843();
+    test_gh_4011();
 }

